### PR TITLE
[SYCLomatic] Add migration of 11 Math APIs; Refine migration of 4 Math APIs by using sycl API, 5 Math APIs by using help function.

### PR DESCRIPTION
**Pull request by tangjj11**
Signed-off-by: Tang, Jiajun jiajun.tang@intel.com

### CODE REVIEW

**Comment by zjin-lcf** [comment]
Hello, I have some questions after taking a quick look. Are these extended math functions supported by the compiler? Is there a table for these functions? Thanks.

**Comment by tangjj11** [comment]
Hi, @zjin-lcf. These extended math API are supported by the compiler. You can use them by include <sycl/ext/intel/math.hpp>. And you can find all the extended math API list in this file.

**Comment by zjin-lcf** [comment]
Is this the file you referred to  ?  
https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/math.hpp
For example, normcdf is not there. Right ?

**Comment by tangjj11** [comment]
@zjin-lcf Yes, I referred to this file. But we have a corresponding internal sycl compiler repo, and there are more API than this file (https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/math.hpp). The sycl compiler release version is compiled from internal repo, so we can use them here. You can find these API in next oneAPI release version.

**Review by zhiweij1 — APPROVED** [review]

**Review by ziranzha — APPROVED** [review]

File: clang/runtime/dpct-rt/include/math.hpp.inc [hunk 1]
```diff
@@ -42,6 +42,62 @@
...
+  case 2:
+    return sycl::length(sycl::vec<T, 2>(a[0], a[1]));
+  case 3:
+    return sycl::length(sycl::vec<T, 3>(a[0], a[1], a[2]));
+  case 4:
+    return sycl::length(sycl::vec<T, 4>(a[0], a[1], a[2], a[3]));
+  default:
+    T ret = a[0];
+    for (int i = 1; i < len; ++i)
+      ret = sycl::hypot(ret, a[i]);
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/math.hpp.inc]: let's add the num 1st,  then do sqrt.

File: clang/runtime/dpct-rt/include/math.hpp.inc [hunk 2]
```diff
@@ -42,6 +42,62 @@
...
+
+// DPCT_LABEL_BEGIN|length|dpct
+// DPCT_DEPENDENCY_EMPTY
+// DPCT_CODE
+/// Calculate the square root of the input array.
+/// \param [in] a The array pointer
+/// \param [in] len Length of the array
+/// \returns The square root
+template <typename T> inline T length(const T *a, const int len) {
+  if (len < 1)
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/math.hpp.inc:81]: how about put if in default, then when 1, 2, 3,4, it don't need exe this branch.

Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/math.hpp.inc]: suggest to remove the if statment, ref to fast_length()

File: clang/lib/DPCT/APINames.inc [hunk 3]
```diff
@@ -620,20 +620,20 @@ ENTRY(modff, modff, true, NO_FLAG, P0, "Successful: DPCT1017")
...
-ENTRY(norm3df, norm3df, true, NO_FLAG, P4, "DPCT1017")
-ENTRY(norm4df, norm4df, true, NO_FLAG, P4, "DPCT1017")
-ENTRY(normcdff, normcdff, false, NO_FLAG, P4, "Unsupported: [TODO]: explain why unsupported")
-ENTRY(normcdfinvf, normcdfinvf, false, NO_FLAG, P4, "Unsupported: [TODO]: explain why unsupported")
-ENTRY(normf, normf, false, NO_FLAG, P0, "Unsupported: [TODO]: explain why unsupported")
+ENTRY(norm3df, norm3df, true, NO_FLAG, P4, "Successful")
+ENTRY(norm4df, norm4df, true, NO_FLAG, P4, "Successful")
+ENTRY(normcdff, normcdff, true, NO_FLAG, P4, "Successful")
+ENTRY(normcdfinvf, normcdfinvf, true, NO_FLAG, P4, "Successful")
+ENTRY(normf, normf, true, NO_FLAG, P0, "Successful")
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APINames.inc]: Add DPCT1017 if there is precision issue.

Comment (tangjj11) [reply clang/lib/DPCT/APINames.inc]: fixed

File: clang/runtime/dpct-rt/include/math.hpp.inc (hunk 2 above)
Comment (tangjj11) [reply clang/runtime/dpct-rt/include/math.hpp.inc]: fixed

File: clang/runtime/dpct-rt/include/math.hpp.inc (hunk 1 above)
Comment (tangjj11) [reply clang/runtime/dpct-rt/include/math.hpp.inc]: fixed

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**